model.py

```diff
@@ -137,3 +192,246 @@ def forward(self, inputs: Dict[str, th.Tensor]) -> th.Tensor:
+                             'i o -> b k i o', b=rxn.shape[0], k=1)
+        rxn = th.cat([rxn, rxn0], dim=1)
+        piv0 = th.zeros_like(piv[:, :1])
+        piv = th.cat([piv, piv0], dim=1).unsqueeze(-1)
+
+        xyz = einops.repeat(self.xyz, 'd x y z -> b k d (x y z)',
+                            b=rxn.shape[0], k=K)
+        xyz2 = rxn @ (xyz - piv) + piv + txn
+        motion = (xyz2 - xyz).view(*xyz.shape[:3], *self.xyz.shape[-3:])
+        motion = th.sum(motion * th.unsqueeze(clf, 2), 1)
+        return motion
+
+
+class SO3(nn.Module):
+    """Euler angle parameterization of the rotation matrix."""
+
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, x: th.Tensor) -> th.Tensor:
+        """Rotation matrix from euler angle parameterization.
+
+        Args:
+            x: Euler angles of shape (..., 3).
+                (roll,pitch,yaw) convention.
+        Returns:
+            Rotation matrix of shape (..., 3, 3),
+                R@x convention.
+        """
+
+        c, s = th.cos(x), th.sin(x)
+        cx, cy, cz = [c[..., i] for i in range(3)]
+        sx, sy, sz = [s[..., i] for i in range(3)]
+
+        # NOTE(ycho):
+        # Represents Rz@Ry@Rx matrix,
+        # Which is equivalent to:
+        # rotate_yaw(rotate_pitch(rotate_roll(point))).
+        r = th.stack([
+            cy * cz,
+            (sx * sy * cz) - (cx * sz),
+            (cx * sy * cz) + (sx * sz),
+            cy * sz,
+            (sx * sy * sz) + (cx * cz),
+            (cx * sy * sz) - (sx * cz),
+            -sy,
+            (sx * cy),
+            (cx * cy)], dim=-1)
+
+        r = einops.rearrange(r, '... (i o) -> ... i o', i=3, o=3)
+        return r
+
+
+class MotionPredictor(nn.Module):
+    """DSR-Net Motion Predictor.
+
+    Predicts scene flow from scene representation and action embeddings.
+    Uses euler angle parameterization internally.
+    """
+
+    def __init__(self, use_action: bool, num_objects: int = 5,
+                 num_params: int = 6):
+        """
+        Args:
+            use_action: Whether `action` embeddings will be provided as input.
+                Otherwise, will generate a zero-filled tensor with same shape.
+            num_objects: Maximum number of object instances to predict.
+            num_params: The number of parameters for the transform encoding.
+                By default, uses 6 = (3 for translation, 3 for rotation).
+        """
+        super().__init__()
+        # NOTE(ycho): `use_action` is only `False`
+        # while we don't have ActionEmbedding() implemented.
+        self.use_action: bool = use_action
+        self.num_objects: int = num_objects
+        self.num_params: int = num_params
+
+        self.conv3d0 = Conv3D(8 + 8, 8, 2)
+        self.conv3d1 = Conv3D(8 + 8, 16, 2)
+        self.conv3d2 = nn.Sequential(*[
+            Conv3D(i, o, s) for
+            (i, o), s in zip(pairwise(
+                [16 + 16, 32, 32, 32, 64]), (2, 1, 1, 1))])
```

Comment:
Not a serious issue, but I wonder why you used `zip(pairwise(channels), strides)` here while `epairwise(channels)` was used in `SceneEncoder`.